code/tracing/tracing.cpp

```diff
@@ -103,507 +70,299 @@ static int get_query_object() {
-{
-	if (Cmdline_json_profiling)
-	{
-		std::lock_guard<std::mutex> guard(json_mutex);
-
-		tracing_data data;
-
-		data.name = name;
-		
-		data.pid = get_pid();
-		data.tid = get_tid();
-
-		data.enter = true;
-		data.time = timer_get_nanoseconds();
-
-		if (do_gpu_queries) {
-			data.gpu_query = get_query_object();
-			gr_query_value(data.gpu_query, QueryType::Timestamp);
-		} else {
-			data.gpu_query = -1;
-		}
-
-		current_frame_data.push_back(data);
+	if (do_gpu_queries) {
+		gpu_start_query = get_gpu_timestamp_query();
 	}
+	cpu_start_time = timer_get_nanoseconds();
 
-	if (Cmdline_frame_profile)
-	{
-		int parent = -1;
-		for (int i = 0; i < (int)samples.size(); i++) {
-			if ( !samples[i].open_profiles ) {
-				continue;
-			}
-
-			samples[i].num_children++;
-
-			if (samples[i].num_children == 1) {
-				// this is our direct parent for this new sample
-				parent = i;
-			}
-		}
-
-		for(int i = 0; i < (int)samples.size(); i++) {
-			if( !strcmp(samples[i].name.c_str(), name) && samples[i].parent == parent ) {
-				// found the profile sample
-				samples[i].open_profiles++;
-				samples[i].profile_instances++;
-				samples[i].start_time = timer_get_microseconds();
-				Assert(samples[i].open_profiles == 1); // max 1 open at once
-				return;
-			}
-		}
-
-		// create a new profile sample
-		profile_sample new_sample;
+	main_thread_id = get_tid();
 
-		new_sample.name = SCP_string(name);
-		new_sample.open_profiles = 1;
-		new_sample.profile_instances = 1;
-		new_sample.accumulator = 0;
-		new_sample.start_time = timer_get_microseconds();
-		new_sample.children_sample_time = 0;
-		new_sample.num_children = 0;
-		new_sample.parent = parent;
+	initialized = true;
+}
 
-		samples.push_back(new_sample);
+void process_events() {
+	if (do_gpu_queries) {
+		// Process pending GPU events
+		process_gpu_events();
 	}
 }
 
-/**
- * Used to end profiling of a section of code. Note that the parameter given MUST match that of the preceding call
- * to profile_begin
- * @param name A globally unique string that will be displayed in the HUD readout
- */
-void profile_end(const char* name)
-{
-	if (Cmdline_json_profiling)
-	{
-		std::lock_guard<std::mutex> guard(json_mutex);
+void shutdown() {
+	while (!gpu_events.empty()) {
+		process_events();
 
-		tracing_data data;
+		// Don't do busy waiting...
+		os_sleep(5);
+	}
 
-		data.name = name;
+	for (auto query : query_objects) {
+		gr_delete_query_object(query);
+	}
+	query_objects.clear();
 
-		data.pid = get_pid();
-		data.tid = get_tid();
+	mainFrameTimer = nullptr;
+	traceEventWriter = nullptr;
 
-		data.enter = false;
-		data.time = timer_get_nanoseconds();
+	initialized = false;
+}
 
-		if (do_gpu_queries) {
-			data.gpu_query = get_query_object();
-			gr_query_value(data.gpu_query, QueryType::Timestamp);
-		} else {
-			data.gpu_query = -1;
-		}
+namespace complete {
 
-		current_frame_data.push_back(data);
+void start(const Category& category, trace_event* evt) {
```

Comment:
That's exactly what is being done. The PID isn't hugely relevant since we only use one process but if we ever use more than one this should be able to handle that. The TID is correctly recorded and I already tested it with the new multithreaded video player.